### DISCUSSION TITLE
New Variables {{caret_content}} related

### DISCUSSION THREAD

**Discussion by FelipeRearden** [Taitava/obsidian-shellcommands#117] [Ideas]
Hello @Taitava 

Following our conversations about caret position in #116 and #114 

New variables to select the content of inside the caret position in the current file. It’s a new variable that selects text using the caret position as a method of search 

{{caret_content:line}} - get the content of an entire line

{{caret_content:line:column}} - get the WORD that is inside of a specific caret position  

{{caret_content:current}} - get the WORD that is inside the current caret position 

Example 1: get the content of the line 11 in the current file and send to an specific note INP.md

Echo {{caret_content:11}} >> INP.md

Example 2: send the WORD to my note of Swedish words that I need to learn 

Echo {{caret_content:current}} >> Swedish_Dictionary.md

Have a great day!

**Comment by Taitava** [comment]
> {{caret_content:line}} - get the content of an entire line
> 
> {{caret_content:line:column}} - get the WORD that is inside of a specific caret position

I'd change the naming here. When referring to a certain line and/or column, there's no need to access the caret's position, so there's also no need to include the word _caret_ in the variable's name.

> {{caret_content:current}} - get the WORD that is inside the current caret position

This instead is related to the caret, so the word _caret_ is good here. I'd just change the argument `current` to `word` to be more descriptive: `{{caret_content:word}}`. Then we can add more: sentence, line, paragraph, chapter, or everything under the current heading (with/without the heading), etc.

## Another way: Use distinct variables like `{{word}}`, `{{line}}`, `{{sentence}}` etc.
These variables would not have `caret` in their name at all
Parameters for each of these variables:
- `:position`: line/column. (if not defined, use caret position).
- `:separators`: What characters to use as word boundaries / sentence boundaries.
- Variable specific parameters, e.g. `{{line}}` could have a parameter that defines whether a trailing newline character should be included.


> Example 2: send the WORD to my note of Swedish words that I need to learn
> 
> Echo {{caret_content:current}} >> Swedish_Dictionary.md

Thanks, that's a nice reference to our other conversation! 😎

**Reply by Taitava** [reply]
## Multicaret support

This adds on top of the _Another way_ idea:
- Have the variables also in plural form: `{{words}}`, `{{lines}}`, `{{sentences}}`
- These do not have a `:position` parameter, they always use carets' positions.
- They need another `:separator` parameter for separating the content that comes from multiple carets.
- Normal, singular named `{{word}}`, `{{line}}`, `{{sentence}}` are also available.
- I think I had something more on my mind, but I forgot it.

**Reply by FelipeRearden** [reply]
Fantastic @Taitava !!! 

Great modifications, you always bring the right concept and terms to my ideas :) 

Now its better and much more easy to understand :) 

> Then we can add more: sentence, line, paragraph, chapter, or everything under the current heading (with/without the heading), etc.

👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 👀 

> Another way: Use distinct variables like {{word}}, {{line}}, {{sentence}} etc.
>
> These variables would not have caret in their name at all
Parameters for each of these variables:

This is gonna be another huge feature @Taitava !!!!!!!

The amount of text manipulation/transformation that we will be able to do is amazing !!!!!

> Thanks, that's a nice reference to our other conversation! 😎

I knew you would like 😃 😃 😃 😃 😃 😃 😃 😃

**Reply by FelipeRearden** [reply]
> Multicaret support
> 
> This adds on top of the Another way idea:
> 
> Have the variables also in plural form: {{words}}, {{lines}}, {{sentences}}
> These do not have a :position parameter, they always use carets' positions.
> They need another :separator parameter for separating the content that comes from multiple carets.
> Normal, singular named {{word}}, {{line}}, {{sentence}} are also available.
> I think I had something more on my mind, but I forgot it.

AMAZING !!!!!!!!!!!!!!!!! 🤯 🤯 🤯 🤯 🤯 🤯 🤯 🤯 🤯 🤯 

@Taitava we need this variable to work with `output=clipboard` 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 #86

**Reply by Taitava** [reply]
Ah, I just realised that the terms _words_, _sentences_, _lines_ etc can have a different meaning, too: They can refer to **all** words/sentences/lines in the current note, e.g. _get **all** lines and separate them with a comma_, regardless of where carets are positioned. To overcome this ambiguity, should the caret related variables be:

- `{{caret_word}}`
- `{{caret_words}}`
- `{{caret_sentence}}`
- `{{caret_sentences}}`
- ~~`{{caret_line}}` (aaah, can be confused with position, although here it means a text line)~~ Edit 2022-11-03: It will be  `{{caret_paragraph}}`. Edit 2022-11-26: This is now released.
- ~~`{{caret_lines}}`~~ `{{caret_paragraphs}}`
- etc
?

A good thing with this naming convention is that then these variables, that are quite similar to each other, would appear together in variable lists.

**Reply by Taitava** [reply]
Maybe replace `{{caret_line}}` with `{{caret_paragraph}}`? Obsidian has a command named _Delete paragraph_, which basically deletes the current line.

**Reply by Taitava** [reply]
I've decided to use `{{caret_paragraph}}`.

**Reply by Taitava** [reply]
`{{caret_paragraph}}` will be the first one of these new variables. I've made an issue for it: https://github.com/Taitava/obsidian-shellcommands/issues/282

**Reply by Taitava** [reply]
`{{caret_paragraph}}` is now released in `0.17.0`.
